ARRBaseRobot Add CreateRobotROS2Interface()

Tadinu opened the pull request:
* RobotVehicleROSController: ARobotVehicle -> ARRRobotBaseVehicle
* Add ARRCrowAIController due to [ADetourCrowdAIController](https://github.com/EpicGames/UnrealEngine/blob/release/Engine/Source/Runtime/AIModule/Classes/DetourCrowdAIController.h#L12) lacks module API tag, thus not compilable in EDITOR build
* `ARRBaseRobot` add ROS2Interface member & `CreateRobotROS2Interface()`, only instantiating it upon valid ROS2InterfaceClass
* Add Scripts/verify_ue4_env.sh, so users could use a common custom env var (as path to UE4Editor) for all running scripts in client project repos
* `URRRobotROS2Interface` Add `OnMessageReceived()`, being used as a generic class callback, eg [here](https://github.com/rapyuta-robotics/RapyutaSimulationInternal/blob/SOOTBALL_SIM/Source/RapyutaSimInternal/Private/Sootball/RRSootballROS2Interface.cpp#L109).

Commit:
URRRobotROS2Interface Add OnMessageReceived()
Robot: ARobotVehicle -> ARRBaseRobot

Commit:
ARRBaseRobot add ROS2Interface member

Commit:
RRBaseRobot add CreateRobotROS2Interface
* Only instantitate it upon valid ROS2InterfaceClass
Add Scripts/verify_ue4_env.sh

Commit:
minor refactoring

yuokamoto commented:
it is better to rename pr. should not use sootballs in public repo.

Commit:
0.0.10 Changelog

tetorea commented on Source/RapyutaSimulationPlugins/Private/Robots/RRBaseRobot.cpp:
```diff
@@ -28,9 +29,37 @@ void ARRBaseRobot::SetupDefault()
+    URRUObjectUtils::SetupDefaultRootComponent(this);
 
-    RootComponent = CreateDefaultSubobject<USceneComponent>(TEXT("Root"));
-    ROS2InterfaceClass = URRRobotROS2Interface::StaticClass();
```
why is the default initialization of `ROS2InterfaceClass` removed?

Tadinu replied:
Yeah, it is noted [here](https://github.com/rapyuta-robotics/RapyutaSimulationPlugins/blob/SootballUtils2/Source/RapyutaSimulationPlugins/Private/Robots/RRBaseRobot.cpp#L34).
I also don't actually favor leaving it null here, but since child BP class may reconfig this class, creating another instance in `PostInitializeComponents()`.
So basically, `ARRBaseRobot` should not be instantiated directly. This is actually very similar to `ARRBaseRobotVehicle::VehicleComponentClass`, which is reconfigured by child BPs classes like `BP_TurtlebotBurgerPhysics/BP_TurtlebotWaffleVehicle`

tetorea commented on Source/RapyutaSimulationPlugins/Private/Robots/RRBaseRobot.cpp:
```diff
@@ -28,9 +29,37 @@ void ARRBaseRobot::SetupDefault()
+
+void ARRBaseRobot::CreateROS2Interface()
+{
+    ROS2Interface = CastChecked<URRRobotROS2Interface>(
```
Can the `CastChecked` make an exception and crash the editor if `ROS2InterfaceClass` is not initialized properly?

Tadinu replied:
Yes, all child classes (either cpp or BP) of ARRBaseRobot is expected to config it.

tetorea commented on Source/RapyutaSimulationPlugins/Private/Robots/RRRobotROS2Interface.cpp:
```diff
@@ -67,7 +69,8 @@ bool URRRobotROS2Interface::InitPublishers()
         OdomPublisher->InitializeWithROS2(RobotROS2Node);
-        OdomPublisher->RobotVehicle = Robot;
+        // If publishing odom, it must be an [ARRRobotBaseVehicle]
+        OdomPublisher->RobotVehicle = CastChecked<ARRRobotBaseVehicle>(Robot);
```
Same question as above : can this make the editor crash if `Robot` is not properly initialized?

Tadinu replied:
Yes. `CastChecked` acts like a type check for the owner `Robot` if turning on `bPublishOdom`

tetorea commented on Source/RapyutaSimulationPlugins/Public/Core/RRCrowdAIController.h:
```diff
@@ -0,0 +1,19 @@
+#include "RRCrowdAIController.generated.h"
+
+UCLASS()
+class RAPYUTASIMULATIONPLUGINS_API ARRCrowdAIController : public AAIController
```
Is the class `ARRCrowdAIController` used somewhere?

Tadinu replied:
yeah, it's part of [Detour Crowd Manager](https://docs.unrealengine.com/4.27/en-US/InteractiveExperiences/ArtificialIntelligence/NavigationSystem/Avoidance), which works with Pawn+Navmesh & does not require CharacterMoveComponent as with the default RVO.
It is configured here for [RRSootball](https://github.com/rapyuta-robotics/RapyutaSimulationInternal/blob/SOOTBALL_SIM/Source/RapyutaSimInternal/Private/Sootball/RRSootball.cpp#L46)

yuokamoto replied:
it is better to have separate pr to avoid confusion from next time.

yuokamoto commented on Source/RapyutaSimulationPlugins/Public/Robots/RRRobotROS2Interface.h:
```diff
@@ -123,4 +123,30 @@ class RAPYUTASIMULATIONPLUGINS_API URRRobotROS2Interface : public UObject
+             typename TROS2MessageData,
+             typename TRobot,
+             typename TRobotMemFuncType = void (TRobot::*)(const TROS2MessageData&)>
+    FORCEINLINE void OnMessageReceived(const TROS2Message* InMsg, const TRobotMemFuncType& InMemFunc)
```
whew is this used?

Tadinu replied:
noted in the description

yuokamoto reviewed (commented):
Are you able to control robot by publishing cmd_vel from terminal?

Tadinu commented:
> Are you able to control robot by publishing cmd_vel from terminal?
`cmd_vel` is still enabled but only for robot vehicle, which could move around, ARRBaseRobot is mostly an abstract generic class.

yuokamoto commented:
> > Are you able to control robot by publishing cmd_vel from terminal?
> > `cmd_vel` is still enabled but only for robot vehicle, which could move around, ARRBaseRobot is mostly an abstract generic class.

Commit:
ARRBaseRobot::PostInitializeComponents()
Post-call Super:: for AIController to be spawned after ROS2Interface is created
